subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/DefaultPropertyWalker.java

```diff
@@ -190,4 +219,30 @@ public void validate(String propertyName, boolean optional, ValidationAction val
             }
         }
     }
+
+    private static class ImplementationPropertyValue implements ValidatingValue {
+
+        private final HashCode value;
+
+        public ImplementationPropertyValue(Object bean, ClassLoaderHierarchyHasher classLoaderHierarchyHasher) {
+            BuildCacheHasher buildCacheHasher = new DefaultBuildCacheHasher();
+            HashCode classLoaderHash = classLoaderHierarchyHasher.getClassLoaderHash(bean.getClass().getClassLoader());
+            String typeName = bean.getClass().getName();
+            if (classLoaderHash != null) {
```

Comment:
+1 to making the whole cache key for the task invalid if we cannot cache a classloader. We should also make the task out of date every time this happens.
Using `null` makes things worse, since then even changing the class name would not result in the task being out of date, right?

Maybe we should collect the nested beans in a separate list? Like `TaskProperties.nestedBeans`? Then we could use this information for change detection and calculating the build cache key.
WDYT?

Reply:
The implementation hash will now be `null` when the classloader hash is `null`.

Reply:
I implemented the changes making the build cache key for the task invalid.